nucypher/utilities/porter/control/specifications/fields/retrieve.py

```diff
@@ -0,0 +1,53 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+from marshmallow import fields
+
+from nucypher.control.specifications.base import BaseSchema
+from nucypher.control.specifications.exceptions import InvalidInputData
+from nucypher.control.specifications.fields import Base64BytesRepresentation
+from nucypher.crypto.umbral_adapter import CapsuleFrag as CapsuleFragClass
+from nucypher.policy.kits import RetrievalKit as RetrievalKitClass
+from nucypher.utilities.porter.control.specifications.fields import UrsulaChecksumAddress
+
+
+# TODO should this be moved to character control - would this be used by a Bob character control endpoint?
+class RetrievalKit(Base64BytesRepresentation):
+    def _deserialize(self, value, attr, data, **kwargs):
```

Comment:
@fjarri instead of serializing/deserializing based on the bytes representation of a `RetrievalKit` object, should we expect to receive the broken down representation of a RetrievalKit from its constituent parts i.e. serialized to simpler representations and deserialize can create a RetrievalKit object from those values.

Instead of
```
'retrivalkit': <base64_encoded_bytes_of_retrievalkit_object>
```

should it be:
```
'retrieval_kit': {
    'capsule': <base64_encoded_bytes_of_capsule_object>,
    'queried_addresses': [
          <ursula_address_1>,
          <ursula_address_2>,
          ...
     ]
}
```

The latter would remove the need for a RetrievalKit object and bytes representation the web/js layer.

Thoughts?

Reply:
I guess it makes sense for `RetrievalKit` since nobody would probably want to serialize it into bytes. I just wish we had schemas linked to the objects and not halfway across the codebase, like I did in [this gist](https://gist.github.com/fjarri/35dfee392699eb67290c70a00ac86dbd)

Reply:
Also JSON would make it easier to serialize errors

Reply:
Thought a bit more about the update to the structure mentioned above (which I updated to be more specific of what the structure actually looks like). While it works well for JSON, I think the nested structure doesn't work for URL query parameters which Porter currently supports (https://github.com/nucypher/nucypher/pull/2746). It would need a flatter structure to work for URL query parameters.

Possible options could include:
1. Keep the original structure of the list of base64 encoded retrieval kits bytes and document the expected byte structure - currently the serialized byte structure for a retrieval_kit is `<bytes of capsule><byte of ursula1 address><byte of ursula2 address>...`
2. Modify the API to not take any queried_addresses, and instead just the capsule for re-encryption i.e. I think (?) that means that any retries for the same capsule will be for all of the cfrags. Example (if we do this, would probably rename `retrieval_kits` to just `capsules`):
```
'retrival_kits': [
     <base64_encoded_bytes_of_capsule_object>,
     <base64_encoded_bytes_of_capsule2_object>
     ...
]
```
...
(other alternatives?)

Thoughts?

Reply:
> I just wish we had schemas linked to the objects and not located halfway across the codebase, like I did in this gist

I could see that being pretty beneficial. We could address this in a separate PR once the fields are finalized.

Reply:
It kind of depends on whether the users plan to do multi-step retrievals. It is tempting to just send capsules, I admit.

Reply:
@KPrasch any thoughts here on the structure of `retrieval_kits` with respect to multi-step (if retrying, allows for only retrying for reencryptions that failed - perhaps ursula was temporarily unavailable) vs single-step (if retrying, redo all reencryptions irrespective of previous successes) retrievals?

Reply:
Sticking with the ability to retry reencryptions without repeating already successful reencryptions. i.e. option #1.

> Keep the original structure of the list of base64 encoded retrieval kits bytes and document the expected byte structure - currently the serialized byte structure for a RetrievalKit object is `<bytes of capsule><byte of ursula1 address><byte of ursula2 address>` ...